Feature/se 197 typechecker in cpp client

angee opened the pull request:
### Validator for MiniZinc files in C++ gRPC client

This feature adds: 

 + `minizinc_validator.h` a parser and typechecker for MiniZinc files that takes a problem and data file as input. This way errors in MiniZinc files can be spotted before sending the file to the server.

+ `mzn-cli.h` a special command line interface that supports the main command line options that FlatZinc solvers take, to be compatible with the MiniZinc IDE.

 + `mzn-client.cpp` a special MiniZinc client  that uses the MiniZinc validator and the MiniZinc command line interface, and that can be used within the MiniZinc IDE as a "solver".

+ support for connecting to the SolveEngine staging server through flag `--url`

The feature has been successfully tested on Ubuntu on the command line and as "solver" within the MiniZinc IDE.

beatricezal commented:
This repository is public. So I think stage server URL have to be not mentioned in this repository. 

You have localhostFlag and stagingServerFlag, I think you should have only host flag (with default value "solve.satalia.com:443"). This was mention in #1.

angee commented:
@beatricezal I removed the staging server as option and added a command line flag that allows to set the server manually (with the default of "solve.satalia.com:443") as you suggested.

tunnuz commented on cpp/include/cli.h:
```diff
@@ -0,0 +1,78 @@
+
+ protected:
+  /// the problem file, optionally including its path
+  std::string problemFile;
```
Fields and method names don't follow the Google style guidelines, but I just realised I don't follow them precisely either :dango:

angee replied:
I keep them as they are for now - maybe we can agree on a style sometime in the near future to be consistent?

tunnuz commented on cpp/include/cli.h:
```diff
@@ -0,0 +1,78 @@
+   * @param ending ending of a string
+   * @return true if fullString ends with ending
+   */
+  bool hasEnding(std::string const &fullString, std::string const &ending) const;
```
In some other files you follow the convention `[const] type [reference] identifier` (E.g., in the MiniZinc validator)., here you went `type [const] [reference] identifier`) not sure if it's intentional or not. As far as I know the first is recommended.

tunnuz commented on cpp/include/mzn-cli.h:
```diff
@@ -0,0 +1,37 @@
+
+#include <vector>
+
+#include "cli.h"
```
Maybe I forgot to mention this, usually, for a given library `mylib` I use the following structure

```
src
include/mylib
README.md
bin
lib
```

ad the minimum (`include/mylib` allows you to import the headers using `#include "mylib/some_header.hh"`). In the C++ flags I then put `-Iinclude`). It's up to you to follow this, but the code advantage if that it's always clear which library the headers come from.

tunnuz commented on cpp/src/cli.cpp:
```diff
@@ -1,16 +1,22 @@
 #include <string>
 
-#include "../include/simple_client/cli.h"
+#include "../include/cli.h"
```
Same as above for the directory structure.

tunnuz commented on cpp/src/cli.cpp:
```diff
@@ -1,16 +1,22 @@
-CLI::CLI() {}
+CLI::CLI() {
+  dataFlag = "--data";
+  helpMessage = "usage: client " + problemFlag + " <problemfile> "
```
This is most likely not a performance issue, but generally to build a string it's better to use a `std::stringstream` (it allocates memory less often than using `operator+`).

tunnuz commented on cpp/include/minizinc_validator.h:
```diff
@@ -0,0 +1,49 @@
+
+ private:
+  /// MiniZinc's standard library directory
+  std::string _std_lib_dir;
```
Here you use under_scores in one identifier, camelCase in the other.

angee commented on README.md:
```diff
@@ -43,25 +47,65 @@ The grpc examples should come with precompiled client code, or a script for comp
+ The MiniZinc client is a simple gRPC client that also performs parsing and typechecking of MiniZinc files
+ before sending them to the SolveEngine. This way error in MiniZinc models are handled earlier. 
+ 
+ Change into the `cpp/` directory of the repository, and run `make mzn-client`. This will compile and link the 
```
there is a `mzn-client` target in the Makefile, so you can build it.

angee commented on README.md:
```diff
@@ -43,25 +47,65 @@ The grpc examples should come with precompiled client code, or a script for comp
  * the client can be executed by typing `./client` in the `cpp/` directory after
  it has been compiled. As an example, execute
  
  `./client --api-key <your-key> --problem ../problems/minizinc/optimize.mzn`
```
It does not work with MiniZinc yet, because the main server at solve.satalia.com has not yet released the MiniZinc extension. It is possible only in the staging server.

Thanks for mentioning the issue with the `.cnf` file. this sounds like a problem on the SolveEngine's side, I'll investigate it.

angee commented on README.md:
```diff
@@ -43,25 +47,65 @@ The grpc examples should come with precompiled client code, or a script for comp
  `./client --api-key <your-key> --problem ../problems/minizinc/optimize.mzn`
+ 
+ 
+ `./mzn-client --api-key <your-key>  ../problems/minizinc/optimize.mzn`
```
this is the same issue as mentioned above: the standard server does not recognize MiniZinc yet, so you need to use the staging server.

A syntax error is an error from the parser :-)

angee commented:
@tunnuz I just committed fixes for the comments you raised. Could you **rebase and merge** the branch if you are OK with them?
